docs/tutorials/writing-tests.md

```diff
@@ -0,0 +1,170 @@
+---
+breaks: false
+
+description: Writing tests
+
+---
+
+# The tools we use
+
+- [Hardhat](https://hardhat.org/tutorial/): The framework used to run a special Evm used for testing. Read this specific
+  section: [Testing-contracts](https://hardhat.org/tutorial/testing-contracts.html)
+- [ethers-js](https://docs.ethers.io/v5/): A framework used to access the Evm from javascript/typescript and abstract
+  the contracts, so they can be easily called.
+- [Mocha](https://mochajs.org/): The testing framework, you can declare test suites, individual tests and get a report
+  after running them.
+- [Chai](https://www.chaijs.com/): Helpers to make assertions about results inside the tests.
+- [Waffle chai matchers](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html): Some extra chai helpers
+  specific to the blockchain.
+- [Hardhat deploy](https://hardhat.org/guides/deploying.html): This framework let you run some deploy scripts when you
+  run your tests, take snapshots of the Evm and revert to those snapshots when needed.
+
+# Running the tests
+
+There are two different ways to execute the tests and that can generate some confusion. The configuration used by each
+method is different:
+
+1. Hardhat executes mocha: there is a hardhat task that executes mocha and uses the options declared in the mocha
+   section of the `hardhat-config.ts` file.
+2. Mocha executes hardhat: mocha is executed (like in vscode) and when the hardhat library is required an Evm node is
+   run. In this case the configuration is taken from the file `.mocharc.js`
+
+Our project have the two configurations working but some time there can be some differences. You can run a single test
+using your editor, or you can run all the tests by executing `yarn test` in the project root directory.
+
+## A simple test
+
+Check the typical test in: [Testing-contracts](https://hardhat.org/tutorial/testing-contracts.html)
+
+Each test file can contain some related tests, and you can have a lot of files (or test suites) that test different
+contracts.
+
+Something very important to note is: ***the Evm keeps the state between tests***, so each test must do a clean
```

Comment:
Is it really the EVM that keeps the state between tests?

Reply:
The evm is the most significant, if you don't do snapshot/revert the evm keeps minting blocks and everything is there. The most notorious example is the account balances. 

Hardhat-deploy has two extra things: 

- an in memory cache of the deployments
- the deployments in deployments/hardhat, that in some cases that I didn't manage to reproduce get written.